File: DataStructures/Array/csharp/merge-sorted-arrays-app/merge-sorted-arrays.cs
//Given two sorted arrays, can you combine them into one large one that is still sorted?

using System;

class Program
{
  static int[] MergeSortedArrays(int[] array1, int[] array2)
  {
    // Check for edge cases where one or both arrays are empty
    if (array1.Length == 0)
    {
      return array2; // Return the non-empty array
    }

    if (array2.Length == 0)
    {
      return array1; // Return the non-empty array
    }

    // Initialize the merged array with enough space to hold all elements
    int[] mergedArray = new int[array1.Length + array2.Length];

    int i = 0; // Pointer for array1
    int j = 0; // Pointer for array2
    int k = 0; // Pointer for mergedArray

    // Merge the arrays while they both have elements
    while (i < array1.Length && j < array2.Length)
    {
      if (array1[i] < array2[j])
        mergedArray[k++] = array1[i++];
      else
        mergedArray[k++] = array2[j++];
    }

    // If there are remaining elements in array1, copy them to mergedArray
    while (i < array1.Length)
      mergedArray[k++] = array1[i++];

    // If there are remaining elements in arr2ay, copy them to mergedArray
    while (j < array2.Length)
      mergedArray[k++] = array2[j++];

    return mergedArray; // Return the sorted merged array
  }

  static void Main()
  {
    int[] merged = MergeSortedArrays(new int[] { 0, 3, 4, 31 }, new int[] { 4, 6, 30, 34, 67, 77 });
    Console.WriteLine(String.Join(" ", merged));
    // foreach (int item in merged)
    // {
    //   Console.Write(item + " ");
    // }
  }
}

File: DataStructures/Array/javascript/merge-sorted-arrays.js
function mergeSortedArrays(array1, array2) {
  // Check for edge cases where one or both arrays are empty
  if (array1.length === 0 && array2.length !== 0) {
    return array2; // Return the non-empty array
  }
  if (array2.length === 0 && array1.length !== 0) {
    return array1; // Return the non-empty array
  }
  // if (array1.length === 0 && array2.length === 0) {
  //   return []; // Both arrays are empty, return an empty array - This is not needed - one of if statements above will return empty array
  // }

  const mergedArray = [];

  let i = 0; // Pointer for array1
  let j = 0; // Pointer for array2
  let array1Item = array1[i]; // Current element from array1
  let array2Item = array2[j]; // Current element from array2

  while (array1Item || array2Item) {
    // Continue loop until both arrays are fully processed

    if (!array2Item || array1Item < array2Item) {
      // If array2Item is undefined (empty), or array1Item is smaller
      mergedArray.push(array1Item);
      i++;
      array1Item = array1[i]; // Move pointer in array1
    } else {
      mergedArray.push(array2Item);
      j++;
      array2Item = array2[j]; // Move pointer in array2
    }
  }

  return mergedArray; // Return the sorted merged array
}

console.log(mergeSortedArrays([0, 3, 4, 31], [4, 6, 30, 34, 67, 77]));
// If your solution works - explain how you wold clean this code - make smaller methods etc.
